Extend dev-guidelines handbook (#16)
teamdigitale/io-handbook: development-styleguide.md

## 👨‍💻 DEVELOPMENT STYLE GUIDE

> A practical reason to use a style guide is that they help you write code.

Reviewer: ```suggestion
A practical reason to use a style guide is that it helps you write code.
```

> Furthermore, they show commons mistakes or scenario and how to handle them

Reviewer: ```suggestion
Furthermore, it shows commons mistakes or scenarios and how to handle them
```

### Typescript

> #### Object modification. Why you should avoid it.

Reviewer: ```suggestion
#### Avoid object modification.
```

- Due to its dynamic nature, JavaScript makes it extremely easy to modify objects that you do not own
- Such practice should generally be avoided as it can lead to maintenance problems and can produce hard to find bugs
- If you don't own it, don't modify it

side effects, example

```typescript
type Person = {
  name: string;
  surname: string;
  age: number;
};

function printNameMutation(p: Person) {
  console.log(`${p.name} ${p.surname}`);

>   // this object mutation could broken software logic

Reviewer: ```suggestion
  // this object mutation could break software logic
```

  p.name = "Rob";
}

function printName(p: Person) {
  console.log(`${p.name} ${p.surname}`);
}

> const david: Person = { name: "David", surname: "Smith", age: 20 };

Reviewer: this example looks like it has more to do with "use pure functions". I'll replace with an example that does not use functions. ie. 

```ts
let o = { x: 2 }
o.x = 1; // avoid
// use instead
const a = {...};
const b = { ...a, x: 1 }
```

printNameMutation(david); // David Smith
// the object has been mutated
printName(david); // Rob Smith
```

### [FP-TS](https://gcanti.github.io/fp-ts/modules/Option.ts.html)

> #### [Option](https://gcanti.github.io/fp-ts/modules/Option.ts.html)

Reviewer: ```suggestion
#### Use [Options](https://github.com/gcanti/fp-ts/blob/1.12.0/src/Option.ts) to handle nullable values.
```

we need to link fp-ts 1.x examples (not 2.x)

```typescript
import { Option, some } from "fp-ts/lib/Option";

// this is a common scenario where we want to model an user with email preferences

// undefined -> user doesn't give any feedback
// true -> users wants to receive email
// false -> users doesn't want to receive email
type EmailPreference = boolean | undefined;

type User = {
  id: number;
  emailPreference: EmailPreference;
};

const user: User = { id: 1, emailPreference: true };

// check if user wants to receive email
if (user.emailPreference !== undefined && user.emailPreference === true) {
  // send email
} else {
  console.log("can't send email due to the user's preference");
}

// use option to avoid if/else
type UserO = {
  id: number;

>   emailPreference: Option<boolean>;

Reviewer: I'd rather avoid to introduce `UserO`  and just call

`fromNullable(user.emailPreference).map(...)` to replace the if/then part

moreover I'd add a link to this example https://github.com/gcanti/fp-ts/blob/1.12.0/src/Option.ts#L150
which shows how to extract deep nested nullable values

};

const userO: UserO = { id: 1, emailPreference: some(true) };
// send email if user's preference is set to true
userO.emailPreference.fold(
  () => {
    console.log("can't send email due to the user's preference");
  },
  canSendEmail => {
    // send email
  }
);
```

### REACT / REACT NATIVE

#### Avoid potentially inconsistent state update (React)

> 

Reviewer: to keep code sampes (and effort : ) at minum here  I'd add some links where possibile ie.

https://medium.com/react-ecosystem/how-to-handle-state-in-react-6f2d3cd73a0c

For some scenario, you may be interested why a computation failed.
Using _Option_ is useful when you want to know if a value is present or not.
The reason why it is not present it's not interesting.
If you want to know more about computation failure you should use Either.

Consider this example

> write a function to validate a password
>
> it has to be:
>
> length >= 8
>
> contain at least a char in upper case
>
> contain at least a number

#### a possible implementaion
```typescript
// return the error message if the password is not compliant, null otherwise
export const validatePassword = (pwd: string): string | null => {
  if (pwd.length < 8) {
    return "at least 8 characters";
  }
  if (!/[A-Z]/g.test(pwd)) {
    return "at least one capital letter";
  }
  if (!/[0-9]/g.test(pwd)) {
    return "at least one number";
  }
  return null;
};

const myPws = "123456A";

const validatePwd = validatePassword(myPws);
if (myPws === null) {
  console.log("password is ok");
} else {
  console.error(`the password is too weak: ${validatePwd}`);
}
```
#### using Either
```typescript
// return the error if the password is not compliant, the password otherwise

> export const validatePassword = (pwd: string): Either<Error, string> => {

Reviewer: this example is better handled with `Validation`  (accumulating errors),
see https://dev.to/gcanti/getting-started-with-fp-ts-either-vs-validation-5eja

Author: Since **[getValidation](https://github.com/gcanti/fp-ts/blob/master/src/Either.ts)** is available from version 2.0.0/typescript 3.5.+ (almost of our projects are not using the latest version) and I found that example too complex for this scope (validation and semigroup) I arranged the code to a basic example but I left the link for these cases where developers wants handle multiple errors

Do you think could it be enough?

Reviewer: right !

    return left(Error("at least 6 characters"));
  }
  if (!/[A-Z]/g.test(pwd)) {
    return left(Error("at least one capital letter"));
  }
  if (!/[0-9]/g.test(pwd)) {
    return left(Error("at least one number"));
  }
  return right(pwd);
};
const myPws = "123456A";
const maybePassword = validatePassword(myPws);
maybePassword.fold(
  err => {
    console.error(`the password is too weak: ${err.message}`);
  },
  pwd => {
    console.log("password is ok");
  }
);
```

### [ITALIA-TS-COMMONS](https://github.com/teamdigitale/io-ts-commons)

#### Use [Pot](https://github.com/teamdigitale/io-ts-commons/blob/master/src/pot.ts) to handle the states of remote (potential) data.

Usually an application could deal with loading some data, from remote or local sources.
Pot (for potential) is a data type that helps you to keep track of states, value and errors.
This example shows a reducer that handles the loading of a remote user

```typescript
type User = {
  name: string;
  username: string;
};

type UserState = {
  currentUser: pot.Pot<User, Error>;
};

function userReducer(
  state: UserState = { currentUser: pot.none },
  action: Action
): UserState {
  switch (action.type) {

>     case getType(contentServiceLoad.request):

Reviewer: maybe it's worth to introduce typesafe-actions (async) here ?
